### PR TITLE
Start VM in Harvester

### PR DESCRIPTION
**Pull request by mads-hartmann**
## Description

This is the initial step towards providing dedicated VMs for preview environments.

This introduces a new option with-vm for the build Werft job which will start a new VM in our Harvester cluster. This is a tiny step towards providing dedicated k3s clusters for preview environments.

For now it only boots a plain Ubuntu VM, so this is not useable by devs yet as it does not have Kubernetes running inside of it, but it gives us something to iterate on.

## Related Issue(s)

Part of https://github.com/gitpod-io/harvester/issues/7

## How to test

Run your job with `with-vm`, e.g

```sh
werft run github -f -a with-vm=true
```

Then in Harvester you should see the VM [here](https://harvester.gitpod-dev.com/dashboard/c/local/harvester/kubevirt.io.virtualmachine).

From Harvester you can get a Serial Console. You can use that to install NGINX

```sh
sudo apt-get install nginx
```

And then you should be able to reach it at `http://mads-communicate-with-harvester.preview.gitpod-dev.com/` where `mads-communicate-with-harvester` is the name of the branch.

The routing is configured in this [PR](https://github.com/gitpod-io/harvester/pull/8).

## Release Notes

```release-note
NONE
```

## Documentation

N/A

/werft with-vm

### CODE REVIEW

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **loujaybee** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @loujaybee` in a comment when ready.

Associated issue: *#7*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by wulfthimm** [comment]
We need also a service to access the VM.
```
apiVersion: v1
kind: Service
metadata:
  name: proxy
  namespace: preview-ubuntu # I followed the scheme of the ingress-proxy where the namespace is always staging-<BRANCH-NAME>
spec:
  ports:
    - name: ssh
      protocol: TCP
      port: 22
      targetPort: 22
    - name: http
      protocol: TCP
      port: 80
      targetPort: 80
    - name: https
      protocol: TCP
      port: 443
      targetPort: 443
    - name: kube-api
      protocol: TCP
      port: 6443
      targetPort: 6443
  selector:
    harvesterhci.io/vmName: ubuntu-01 # This is only an example. It can be any label
  type: ClusterIP

```

File: .werft/vm/manifests.ts [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+      domain:
+        hostname: ${vmName}
+        machine:
+          type: q35
+        cpu:
+          cores: 1
+          sockets: 1
+          threads: 1
+        devices:
+          inputs:
```
Comment (meysholdt) [inline .werft/vm/manifests.ts:53]: Is this an input we want / need?

File: .werft/vm/manifests.ts [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    metadata:
+      annotations:
+        harvesterhci.io/sshNames: "[]"
+      labels:
+        harvesterhci.io/vmName: ${vmName}
+    spec:
+      domain:
+        hostname: ${vmName}
+        machine:
+          type: q35
```
Comment (meysholdt) [inline .werft/vm/manifests.ts:47]: Can we use native virtualization instead of Qemu?

Comment (mads-hartmann) [reply .werft/vm/manifests.ts:47]: @meysholdt It's not clear to me which machines types are considered "native". I was assuming that Qemu would use hardware virtulization either way when available. Their docs are a [bit sparse](https://kubevirt.io/user-guide/virtual_machines/virtual_hardware/#machine-type) but I'm happy to try any type that you'd like if you have a suggestion ☺️

File: .werft/vm/manifests.ts (hunk 1 above)
Comment (mads-hartmann) [reply .werft/vm/manifests.ts:53]: It was part of the default configuration - let me try to remove it and see if things still work ☺️

Comment (mads-hartmann) [reply .werft/vm/manifests.ts]: Serial Console still works even after removing the input devices ✅

File: .werft/vm/manifests.ts (hunk 2 above)
Comment (meysholdt) [reply .werft/vm/manifests.ts:47]: as discussed offline, q35 seems fine and seems to do hardwave acceleration.

**Comment by roboquat** [comment]
LGTM label has been added.  <details>Git tree hash: 4518754cf0d271a1881eb921db873744f78b1327</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/7252#issuecomment-995835080" title="Approved">meysholdt</a>*
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **loujaybee** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @loujaybee` in a comment when ready.

Associated issue: *#7*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

Approval requirements bypassed by manually added approval.

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/7252#issuecomment-995835080" title="Approved">meysholdt</a>*

Associated issue: *#7*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Pull request merged**